### PR TITLE
Implemented Spherical Coordinate System

### PR DESCRIPTION
**Pull request by arihantparsoya**
This is a very basic implementaiton of spherical coordinate system in vector module.

Implementations so far:
- Dot product in spherical coordinates.
- Improved the `express()` function in functions.py to express a spherical vector in cartesian coordinate.

Please comment and give feedback so I can know if I am heading in the right direction. There is lot of more work to be done.
#### sample program:

``` python
>>> from sympy import *
>>> from sympy.vector import CoordSysSpherical
>>> from sympy.vector.functions import express
>>> B = CoordSysSpherical('B')
>>> x = 3*B.rˆ + 4*B.θˆ + 5*B.φˆ
>>> print(y.dot(y))
9*sin(4)**2*cos(5)**2 + 9*cos(4)**2 + 9*sin(4)**2*sin(5)**2
>>> print(x.magnitude())
sqrt(9*sin(4)**2*cos(5)**2 + 9*cos(4)**2 + 9*sin(4)**2*sin(5)**2)
>>> x.components
{B.rˆ: 3, B.φˆ: 5, B.θˆ: 4}
>>> a = express(x,B)
>>> print(a)
(3*sin(4)*cos(5))*B.i + (3*sin(4)*sin(5))*B.j + (3*cos(4))*B.k
```

### CODE REVIEW

**Comment by parsoyaarihant** [comment]
@asmeurer , can you review this?

File: sympy/vector/deloperator.py [hunk 1]
```diff
@@ -14,13 +15,24 @@ class Del(Basic):
...
+            obj._i, obj._j, obj._k = system.i, system.j, system.k
+            obj._system = system
+            obj._name = system.__str__() + ".delop"
+
+        elif isinstance(system,CoordSysSpherical):
+            obj = super(Del, cls).__new__(cls, system)
+            obj._r, obj._θ, obj._φ = system.r, system.θ, system.φ
+            obj._rˆ, obj._θˆ, obj._φˆ = system.rˆ, system.θˆ, system.φˆ
+            obj._system = system
+            obj._name = system.__str__() + ".delop"
```
Comment (vasdommes) [inline sympy/vector/deloperator.py]: lines 23,26,27 (and their duplicates 30,33,34) can be moved outside if/else clause:

```python
        obj = super(Del, cls).__new__(cls, system)

        if isinstance(system, CoordSysCartesian):
            obj._x, obj._y, obj._z = system.x, system.y, system.z
            obj._i, obj._j, obj._k = system.i, system.j, system.k

        elif isinstance(system, CoordSysSpherical):
            obj._r, obj._θ, obj._φ = system.r, system.θ, system.φ
            obj._rˆ, obj._θˆ, obj._φˆ = system.rˆ, system.θˆ, system.φˆ

        obj._system = system
        obj._name = system.__str__() + ".delop"
```

Also there are formatting issues: whitespace is missed after comma in lines 22 and 29.

**Comment by moorepants** [comment]
Note that there is an open PR with work on this by @Upabjojr: https://github.com/sympy/sympy/pull/9937

It would be best if both of you discussed implementations and maybe try to merge the work.

File: sympy/vector/coordsyssph.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    @property
+    def φˆ(self):
+        return self._φˆ
+
+    @property
+    def r(self):
+        return self._r
+
+    @property
+    def θ(self):
```
Comment (moorepants) [inline sympy/vector/coordsyssph.py]: These should be ascii names so that you can easily type them into the interpreter and scripts.

Comment (parsoyaarihant) [inline sympy/vector/coordsyssph.py]: I thought so too, can u suggest me any other alternatives?

Comment (vasdommes) [inline sympy/vector/coordsyssph.py]: I'd suggest `r`, `theta`, `phi` for scalars and  `e_r`, `e_theta`, `e_phi` for unit vectors.

Comment (asmeurer) [inline sympy/vector/coordsyssph.py]: Also non-ASCII identifiers are only allowed in Python 3, but we have to support Python 2.7 as well.

**Comment by Upabjojr** [comment]
It would be nice to have a single class for all coordinate systems. The action of differential operators (gradient, divergence, curl, laplacian) can be determined by three parameters and proper formulae (somewhere in the discussion there was a reference to it).

I suggest, a class where:

* you get these parameters as optional arguments to the constructor (omitting them would generate a Cartesian coordinate system)
* optional formatting strings for base vectors and base functions (see [sympy.diffgeom.CoordSystem](http://docs.sympy.org/latest/modules/diffgeom.html#sympy.diffgeom.CoordSystem) as an example).
* avoid the name _CoordSystem_, because it has already been taken in _sympy.diffgeom_.

**Comment by asmeurer** [comment]
There was a lot of discussion about this earlier this year when several students submitted GSoC applications to work on coordinate systems (none were accepted unfortunately). It may be worth looking at that (search the mailing list). Doing this in full generality and doing it right is definitely a GSoC project worth of work.

**Comment by Upabjojr** [comment]
https://github.com/sympy/sympy/pull/10074

Maybe this is a good PR to finish. It uses the Lamé parameters.

**Comment by parsoyaarihant** [comment]
@Upabjojr , could you suggest me some reading material for implementation of this idea using Lamé parameters?

**Comment by Upabjojr** [comment]
https://en.wikipedia.org/wiki/Curvilinear_coordinates#Relation_to_Lam.C3.A9_coefficients

I saw some kind of related stuff on Differential Geometry and Lie Groups for Physicists by Marián Fecko, though that book's purpose is just loosely related to the purpose of the vector module in SymPy.

The idea: you get the expression for the metric in the new coordinate system (in x-y-z coordinates it's the identity matrix). By metric I mean the matrix to express dot products in the new coordinate system. The result of the dot product has to agree in all coordinate systems, despite the different representations of vectors.

Once you have the metric, you can get the expressions for the gradient, divergence, curl and laplacian with the elements of the metric (I think only the once in the diagonal of the metric are sufficient, but I should check).

**Comment by czgdp1807** [comment]
Closing this in favour of https://github.com/sympy/sympy/pull/9937

**Pull request closed without merging**